### ISSUE TITLE
getch: exit_handler used async-signal-unsafe function exit()

### ISSUE THREAD

**Issue by elfring** [ugtrain/ugtrain#26]
[The function "exit"](http://pubs.opengroup.org/onlinepubs/9699919799/functions/exit.html "Terminate a process.") does not belong to the list of [async-signal-safe functions](https://www.securecoding.cert.org/confluence/display/c/SIG30-C.+Call+only+asynchronous-safe+functions+within+signal+handlers "Call only asynchronous-safe functions within signal handlers.").
I guess that a different program design will be needed for [your function "exit_handler"](https://github.com/ugtrain/ugtrain/blob/814f0ac3b8412dd82e7f1e34fa0a65ffbd60755c/src/lib/getch_l.c#L28 "Reconsider signal handling implementation.").

**Comment by sriemer** [comment]
The problem is that the terminal has to be restored upon exiting (leave getch() input processing). The atexit() function can only register a function to be called at regular exit. If the tool is terminated by a signal like SIGINT, that doesn't work. In the signal handler it doesn't make much sense to try to call the functions registered with atexit(), right. So exit() is the wrong function. Thanks for the hint!

If we think about [the function "_exit"](http://pubs.opengroup.org/onlinepubs/9699919799/functions/_exit.html) then, that would work but we'll notice that it doesn't flush open streams/fds and closing them in undefined. So we have to do that in the signal handler.

Would you be fine with the sequence: restore_getch(), flush streams/fds, close streams/fds, _exit()?

**Comment by elfring** [comment]
Would you like to consider other software design options?
* [sigwaitinfo()](http://pubs.opengroup.org/onlinepubs/9699919799/functions/sigwaitinfo.html "wait for queued signals")
* [Self-pipe trick](http://cr.yp.to/docs/selfpipe.html "Use case description by Daniel J. Bernstein")

**Comment by sriemer** [comment]
Sure, if these have benefits.
Do you have any small examples for their use?
Or can you name tools which use them in a similar way?
I tried to learn from the `top` command from the `procps` package while implementing the getch() handling.

**Comment by sriemer** [comment]
I've just checked the `top` command from `procps` 3.2.8. They also call exit() from a signal handler. Sequence: signal(SIGINT, end_pgm) -> end_pgm() -> bye_bye() -> exit()

**Comment by elfring** [comment]
>They also call exit() from a signal handler.

There are still some asynchronous-unsafe signal handler implementations running.
Would you like to fix them?

How do you think about to move more event processing out of signal handling contexts?

**Comment by sriemer** [comment]
Sounds good. sigwaitinfo() should do that. I would only fix the tools where I notice the problem while reading the source code. I haven't heard about it myself before you made me aware of this. So let's fix this tool first.

**Comment by elfring** [comment]
Are you willing to dedicate another thread for receiving of notifications by the function "[sigwaitinfo](http://stackoverflow.com/questions/6223191/dealing-with-asynchronous-signals-in-multi-threaded-program "Dealing with asynchronous signals in multi-threaded program")" or "[select](http://evbergen.home.xs4all.nl/unix-signals.html "Avoiding races with Unix signals and select() ")"?

**Comment by sriemer** [comment]
The main thread already uses select() to get notified about input coming from the dynamic memory FIFO and from stdin. See [sleep_sec_unless_input2()](https://github.com/ugtrain/ugtrain/blob/814f0ac3b8412dd82e7f1e34fa0a65ffbd60755c/src/lib/system.h#L140).

Ah okay, so the idea would be that the sigwaitinfo() thread catches the signals and notifies the main thread via a pipe about that. Right?

**Comment by elfring** [comment]
Do you understand the linked information sources already?
* How do you think about to perform inter-thread communication by a dedicated condition variable (while omitting a special signal handler)?
* Are you aware of further software design dependencies for a working "self-pipe trick"?

**Comment by sriemer** [comment]
I've tried some things forth and back. I would have to fork off the main process and read from the pipe in the parent all the time to be able exit the children at any time as well. That change is too big as I want to reduce platform-dependent parts and move them as platform-independent as possible into the lib. So I'd like to go with _exit(). The GNU implementation flushes and closes fds. I've also tried abort() but that one creates a core dump.
